<a href="https://colab.research.google.com/github/chamonregina/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Copy_of_LS_DS_113_Join_and_Reshape_Data_Assignment_ipynbrc.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [9]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-12-04 22:39:53--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.97.29
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.97.29|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’

instacart_online_gr 100%[===================>] 196.03M  98.3MB/s    in 2.0s    

2019-12-04 22:39:55 (98.3 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’ saved [205548478/205548478]



In [10]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [11]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [12]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
import pandas as pd
pd.set_option('display.max_columns', None)

In [14]:
aisles = pd.read_csv('aisles.csv')
print(aisles.shape)
aisles.head()



(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [15]:
departments = pd.read_csv('departments.csv')
print(departments.shape)
departments.head()

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [16]:
order_products__prior = pd.read_csv('order_products__prior.csv')
print(order_products__prior.shape)
order_products__prior.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [17]:
order_products__train = pd.read_csv('order_products__train.csv')
print(order_products__train.shape)
order_products__train.head()

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [18]:
orders= pd.read_csv('orders.csv')
print(orders.shape)
orders.head()

(3421083, 7)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [19]:
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

In [20]:

order_products = pd.concat([order_products__prior, order_products__train])
print(order_products.shape)
order_products.head()

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [21]:
order_products = pd.concat([order_products__prior, order_products__train])

order_products['product_id']

0          33120
1          28985
2           9327
3          45918
4          30035
           ...  
1384612    14233
1384613    35548
1384614    35951
1384615    16953
1384616     4724
Name: product_id, Length: 33819106, dtype: int64

In [22]:
order_products['product_id'].isin([4724, 45918])

0          False
1          False
2          False
3           True
4          False
           ...  
1384612    False
1384613    False
1384614    False
1384615    False
1384616     True
Name: product_id, Length: 33819106, dtype: bool

In [23]:
condition = products['product_id'].isin(order_products['product_id'])

products_subset = products[condition]

products_subset

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49683       49684          Vodka, Triple Distilled, Twist of Vanilla   
49684       49685                 En Croute Roast Hazelnut Cranberry   
49685       49686                                   Artisan Baguette   
49686       49687         Smartblend Healthy Metabolism Dry Cat Food   
49687       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  
0            61             19  
1           104             13  
2            94              7  
3            38              1  
4             5             13  
...         ...            ...  
49683       124              5  
49684        42              1  
49685       112              3  
49686        41              8  
49687        73             11  

[49685 rows x 4 columns]

In [24]:
Tally = pd.merge(order_products, products_subset, on='product_id', how='inner')

Tally

order_id  product_id  add_to_cart_order  reordered  \
0                2       33120                  1          1   
1               26       33120                  5          0   
2              120       33120                 13          0   
3              327       33120                  5          1   
4              390       33120                 28          1   
...            ...         ...                ...        ...   
33819101   2269288       27499                 31          0   
33819102   2507248       27499                  9          0   
33819103   2271346       37703                 15          0   
33819104   3004244       43725                 25          0   
33819105   3376312       36233                  6          0   

                                               product_name  aisle_id  \
0                                        Organic Egg Whites        86   
1                                        Organic Egg Whites        86   
2                                        Organic Egg Whites        86   
3                                        Organic Egg Whites        86   
4                                        Organic Egg Whites        86   
...                                                     ...       ...   
33819101                  Non-Dairy Coconut Seven Layer Bar       100   
33819102                  Non-Dairy Coconut Seven Layer Bar       100   
33819103  Ultra Sun Blossom Liquid 90 loads Fabric Enhan...        75   
33819104                               Sweetart Jelly Beans       100   
33819105                            Water With Electrolytes       100   

          department_id  
0                    16  
1                    16  
2                    16  
3                    16  
4                    16  
...                 ...  
33819101             21  
33819102             21  
33819103             17  
33819104             21  
33819105             21  

[33819106 rows x 7 columns]

In [25]:
Tally = Tally.drop(['reordered', 'add_to_cart_order', 'order_id', 'aisle_id', 'department_id'], axis=1)
Tally

product_id                                       product_name
0              33120                                 Organic Egg Whites
1              33120                                 Organic Egg Whites
2              33120                                 Organic Egg Whites
3              33120                                 Organic Egg Whites
4              33120                                 Organic Egg Whites
...              ...                                                ...
33819101       27499                  Non-Dairy Coconut Seven Layer Bar
33819102       27499                  Non-Dairy Coconut Seven Layer Bar
33819103       37703  Ultra Sun Blossom Liquid 90 loads Fabric Enhan...
33819104       43725                               Sweetart Jelly Beans
33819105       36233                            Water With Electrolytes

[33819106 rows x 2 columns]

In [26]:
### Sorted final merged dataframe
Tally = Tally.sort_values(by=['product_id'])
print(Tally.shape)
Tally.head()

(33819106, 2)


product_id                product_name
23893886           1  Chocolate Sandwich Cookies
23894452           1  Chocolate Sandwich Cookies
23894453           1  Chocolate Sandwich Cookies
23894454           1  Chocolate Sandwich Cookies
23894455           1  Chocolate Sandwich Cookies

In [27]:
###Number of ea. product sold
Tally['product_name'].value_counts().head(10)


Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

In [29]:
table1

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

In [30]:
### Take the row index, and add it as a new column
table1 = table1.reset_index()
table1

index  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

In [31]:
###unique identifier = index(names)
tidy1 = table1.melt(id_vars='index', value_vars=['treatmenta', 'treatmentb'])
tidy1

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [32]:
tidy1 = tidy1.rename(columns={
    'index': 'Name', 
    'variable': 'trt', 
    'value': 'result'
})
tidy1

Name         trt  result
0    John Smith  treatmenta     NaN
1      Jane Doe  treatmenta    16.0
2  Mary Johnson  treatmenta     3.0
3    John Smith  treatmentb     2.0
4      Jane Doe  treatmentb    11.0
5  Mary Johnson  treatmentb     1.0

In [33]:
tidy1.trt = tidy1.trt.str.replace('treatment', '')  ###Table1 >Tidy

tidy1

Name trt  result
0    John Smith   a     NaN
1      Jane Doe   a    16.0
2  Mary Johnson   a     3.0
3    John Smith   b     2.0
4      Jane Doe   b    11.0
5  Mary Johnson   b     1.0

In [34]:
###Tidy . Table1
### index: unique identifier
### columns: What do you want to differentiate the columns in wide format
### values: Where are the numbers at - go in the middle of the wide dataframe
wide = tidy1.pivot_table(index='Name', columns='trt', values='result')
wide

trt              a     b
Name                    
Jane Doe      16.0  11.0
John Smith     NaN   2.0
Mary Johnson   3.0   1.0

In [0]:
table2 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

In [36]:
table2.columns.tolist()

['treatmenta', 'treatmentb']

In [37]:
table2.index.tolist()

['John Smith', 'Jane Doe', 'Mary Johnson']

In [38]:
table2

treatmenta  treatmentb
John Smith           NaN           2
Jane Doe            16.0          11
Mary Johnson         3.0           1

In [39]:
table2 = table2.reset_index()
table2

index  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

In [40]:
tidy2 = table1.melt(id_vars='index', value_vars=['treatmenta', 'treatmentb'])
tidy2

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [41]:
###Table2----Tidy
tidy2 = tidy2.rename(columns={'index': 'Sample Name', 'variable': 'Treatment Type', 'value': 'Result'})
tidy2

Sample Name Treatment Type  Result
0    John Smith     treatmenta     NaN
1      Jane Doe     treatmenta    16.0
2  Mary Johnson     treatmenta     3.0
3    John Smith     treatmentb     2.0
4      Jane Doe     treatmentb    11.0
5  Mary Johnson     treatmentb     1.0

In [42]:
####Table 2 Wide Form (Pivot-Table)
### index: unique identifier
### columns: What do you want to differentiate the columns in wide format
### values: Where are the numbers at - go in the middle of the wide dataframe
Tidy2 = tidy2.pivot_table(index='Sample Name', columns='Treatment Type', values='Result')
Tidy2

Treatment Type  treatmenta  treatmentb
Sample Name                           
Jane Doe              16.0        11.0
John Smith             NaN         2.0
Mary Johnson           3.0         1.0

In [43]:
flights = sns.load_dataset('flights')
flights.head()

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [44]:
###Pivot-Table
flights = flights.pivot_table(index='year', columns='month', values='passengers')
flights

month  January  February  March  April  May  June  July  August  September  \
year                                                                         
1949       112       118    132    129  121   135   148     148        136   
1950       115       126    141    135  125   149   170     170        158   
1951       145       150    178    163  172   178   199     199        184   
1952       171       180    193    181  183   218   230     242        209   
1953       196       196    236    235  229   243   264     272        237   
1954       204       188    235    227  234   264   302     293        259   
1955       242       233    267    269  270   315   364     347        312   
1956       284       277    317    313  318   374   413     405        355   
1957       315       301    356    348  355   422   465     467        404   
1958       340       318    362    348  363   435   491     505        404   
1959       360       342    406    396  420   472   548     559        463   
1960       417       391    419    461  472   535   622     606        508   

month  October  November  December  
year                                
1949       119       104       118  
1950       133       114       140  
1951       162       146       166  
1952       191       172       194  
1953       211       180       201  
1954       229       203       229  
1955       274       237       278  
1956       306       271       306  
1957       347       305       336  
1958       359       310       337  
1959       407       362       405  
1960       461       390       432

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####